share/wake/lib/core/list.wake

```diff
@@ -802,3 +805,76 @@ export def groupBy (cmp: a => a => Boolean) (list: List a): List (List a) =
             (elem, head, rest), tail
         _ = (elem, Nil), acc
     foldr combo Nil sorted
+
+export def fromOption (x: Option a): List a =
+    getOrElse Nil $ omap (_, Nil) x
+
+# TODO: replace at with atT
+export def atT (i: Integer) (list: List a): List a =
+    at i list | fromOption
+
+export def atE (i: Integer): (a => a) => List a => List a =
+    def edit f l =
+      require Pair first (x, follow) = splitAt i l else l
+      first ++ (f x, follow)
+    edit
+
+export def headT: (List a) => (List a) = atT 0
+
+export def headE: (a => a) => List a => List a = atE 0
+
+export def takeE (i: Integer): (a => a) => List a => List a =
+  def edit f l =
+    def Pair first follow = splitAt i l
+    (map f first) ++ follow
+  edit
+
+# NOTE: This is an easier-to-use version of `find`
+# Do we care to keep it?
+export def findFirst (p: a => Boolean): List a => List a =
+    find p _ | fromOption | map getPairFirst
+
+export def findFirstE (p: a => Boolean): (a => a) => List a => List a =
+    def edit f l =
+      require (x, xs) = l
+      if p x then
+        f x, xs
+      else
+        x, edit f xs
+
+    edit
+
+export def filterE (p: a => Boolean): (a => a) => List a => List a =
+    def editIf f x = if p x then f x else x
+    map $ editIf _
+
+export def lookup (eq: a => a => Boolean) (key: a): (List (Pair a b)) => List b =
```

Comment:
I've seen this pattern used a ton of places. It definitely belongs in the standard library: 👍 

I'd propose to have a select function instead of an `eq` function, though.

Maybe?
```
export def lookup (selectFn: a => Boolean): List (Pair a b) => List b =
  def select (Pair key value) = if selectFn key then Some value else None
  mapPartial select
```
... again just trying to eliminate unnecessary loop constructions and make the code more direct.